source/common/stats/allocator_impl.cc

```diff
@@ -25,6 +26,29 @@ const char AllocatorImpl::DecrementToZeroSyncPoint[] = "decrement-zero";
 AllocatorImpl::~AllocatorImpl() {
   ASSERT(counters_.empty());
   ASSERT(gauges_.empty());
+
+  // Move deleted stats into the sets for the ASSERTs in removeFromSetLockHeld to function.
+  for (auto& counter : deleted_counters_) {
+    auto insertion = counters_.insert(counter.get());
+    // Assert that there were no duplicates.
+    ASSERT(insertion.second);
+  }
+  for (auto& gauge : deleted_gauges_) {
+    auto insertion = gauges_.insert(gauge.get());
+    // Assert that there were no duplicates.
+    ASSERT(insertion.second);
+  }
+  for (auto& text_readout : deleted_text_readouts_) {
+    auto insertion = text_readouts_.insert(text_readout.get());
+    // Assert that there were no duplicates.
+    ASSERT(insertion.second);
+  }
+  // Make sure all stats are deleted BEFORE threads are shutdown. This is
+  // because the stat destructors need to lock a mutex, which will lead to
+  // undefined behavior once the threads are destroyed.
```

Comment:
ah I see. Your comment is not quite accurate though, and there's a slightly nicer fix.

The problem is order-of-destruction and I think a better fix is to move the mutex to be above the declarations for all the collections in the structure. Then you don't need these explicit clear()s.

Reply:
I don't know how I missed something that simple :-/

Fixed.